### PR TITLE
batches: manually trigger workspaces preview, improvements

### PR DESCRIPTION
**Pull request by courier-new**
A number of improvements to the "Create" screen centered around the workspaces preview here:

https://user-images.githubusercontent.com/8942601/141378001-61dda8cc-8149-4958-823c-d8b16c0c23c5.mov

- Workspaces preview is manually triggered
  - Prompt is shown whenever valid batch spec code has not been previewed
  - Heavy computation (validation) is still debounced
  - Prominent error messages when anything prevents preview
- Numerous code organizational improvements
  - The mega mutation to submit the batch spec is broken down into smaller, smarter, more readable requests
  - "Islands" of business logic related to handling the editor code, submitting the batch spec, polling for the workspace resolution, etc. are collected into custom hooks
  - More components are broken out of the `WorkspacesPreview`
  - Comments everywhere!!!
- Some visual enhancements
  - Workspaces preview is scrollable
  - Prompt from Figma designs
  - Spacing/layout improvements
  - More storybook stories!

### CODE REVIEW

File: client/web/src/enterprise/batches/create/NewCreateBatchChangePage.tsx [hunk 1]
```diff
@@ -78,184 +48,93 @@ const getNamespaceBatchChangesURL = (namespace: SettingsUserSubject | SettingsOr
...
-            // Don't want to trigger on changes to code, it's just the initial value.
-            // eslint-disable-next-line react-hooks/exhaustive-deps
-            [codeUpdates]
-        )
-    )
+    const preview = useCallback(() => {
+        previewBatchSpec(code)
+            // Mark that the batch spec code on the backend is no longer stale.
+            .then(() => setBatchSpecStale(false))
+            .catch(noop)
```
Comment (eseliger) [inline client/web/src/enterprise/batches/create/NewCreateBatchChangePage.tsx:97]: should we add a todo to handle this error? Or is it ignored intentionally?

File: client/web/src/enterprise/batches/create/yaml-util.ts [hunk 2]
```diff
@@ -237,3 +237,11 @@ export const excludeRepo = (spec: string, repo: string, branch: string): YAMLMan
...
     return removeRepoResult
 }
+
+/**
+ * Simply checks for the presence of an "on: " string within the batch spec YAML. Does not
+ * check validity of the spec.
+ *
+ * @param spec the raw batch spec YAML string
+ */
+export const hasOnStatement = (spec: string): boolean => !!spec.match(/^on:\n\s*- \S+/m)
```
Comment (eseliger) [inline client/web/src/enterprise/batches/create/yaml-util.ts:247]: maybe we can actually parse the spec here to check? it seems a little brittle to do that kind of check, there would be ways this doesn't match (JSON for example, which is valid YAML too)

**Review by eseliger — APPROVED** [review]
So excited to ship this! Great work, very high quality components :)

File: client/web/src/enterprise/batches/create/yaml-util.ts (hunk 2 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/create/yaml-util.ts:247]: JSON is valid YAML? 🤯 Yeah that makes sense; I was thinking the same but just mostly wanted to get  "close enough" for now. I'll add a TODO and address this in a follow up!

File: client/web/src/enterprise/batches/create/NewCreateBatchChangePage.tsx (hunk 1 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/create/NewCreateBatchChangePage.tsx:97]: It's intentionally ignored! The error should have already been surfaced from `previewBatchSpec` itself. But I thought of a nicer way to achieve the same effect of this callback so I think I'm going to get rid of this anyway. 😄

**Pull request merged**